In [40]:
import cohere
import pandas as pd
import numpy as np
import os
from dotenv import load_dotenv
# from sklearn.metrics import classification_report, confusion_matrix
import matplotlib.pyplot as plt
import seaborn as sns
import tqdm

load_dotenv()

co = cohere.Client(os.getenv("COHERE_API_KEY"))

In [11]:
# Convert Data
data_pth = "data/train.txt"
df = pd.read_csv(data_pth, sep=";", header=None, names=['text', 'label'])
df.to_csv("data/transformed.csv", index=False)


single_label_dataset = co.datasets.create(
    name="single-label-dataset",
    data=open("data/transformed.csv", "rb"),
    type="single-label-classification-finetune-input"
)

print(co.wait(single_label_dataset).dataset.validation_status)

from cohere.finetuning.finetuning import(
    BaseModel,
    FinetunedModel,
    Settings,
)

...
...
validated


In [7]:
# List all fine-tuned models
finetuned_models = co.finetuning.list_finetuned_models()

# Print the models
if isinstance(finetuned_models, tuple) and len(finetuned_models) >= 2:
    # Get the list of models from the second element of the tuple
    models_list = finetuned_models[1]

    # Print each model's details
    for model in models_list:
        print(f"ID: {model.id}")
        print(f"Name: {model.name}")
        print(f"Status: {model.status}")
        print(f"Created at: {model.created_at}")
        print("-----------------------------------")
else:
    print("Unexpected response format:", finetuned_models)


Unexpected response format: finetuned_models=[FinetunedModel(id='c62448a1-530a-45bb-8ca4-9e4950f39ac0', name='single-label-finetune', creator_id='89783f58-3842-4032-8f10-9ec9fae0ff55', organization_id='4dcd1a7b-dcae-4e35-a6e6-8c444318f7f0', settings=Settings(base_model=BaseModel(name='english', version='1.0.0', base_type='BASE_TYPE_CLASSIFICATION', strategy='STRATEGY_TFEW'), dataset_id='single-label-dataset-tvf7wf', hyperparameters=None, multi_label=None, wandb=None), status='STATUS_READY', created_at=datetime.datetime(2025, 3, 13, 18, 56, 23, 717399, tzinfo=datetime.timezone.utc), updated_at=datetime.datetime(2025, 3, 13, 19, 12, 59, 738757, tzinfo=datetime.timezone.utc), completed_at=datetime.datetime(2025, 3, 13, 19, 12, 59, 738757, tzinfo=datetime.timezone.utc), last_used=None)] next_page_token=None total_size=1


Unexpected response format: finetuned_models=[FinetunedModel(id='c62448a1-530a-45bb-8ca4-9e4950f39ac0', 
name='single-label-finetune', 
creator_id='89783f58-3842-4032-8f10-9ec9fae0ff55', organization_id='4dcd1a7b-dcae-4e35-a6e6-8c444318f7f0', 
settings=Settings(
    base_model=BaseModel(name='english', 
        version='1.0.0', 
        base_type='BASE_TYPE_CLASSIFICATION', 
        strategy='STRATEGY_TFEW'), 
    dataset_id='single-label-dataset-tvf7wf', 
    hyperparameters=None, 
    multi_label=None, 
    wandb=None), 
    status='STATUS_READY', 
    created_at=datetime.datetime(2025, 3, 13, 18, 56, 23, 717399, tzinfo=datetime.timezone.utc), 
    updated_at=datetime.datetime(2025, 3, 13, 19, 12, 59, 738757, 
        tzinfo=datetime.timezone.utc), 
    completed_at=datetime.datetime(2025, 3, 13, 19, 12, 59, 738757, tzinfo=datetime.timezone.utc), 
    last_used=None)] 

next_page_token=None 
total_size=1

In [8]:
MODEL_ID = 'c62448a1-530a-45bb-8ca4-9e4950f39ac0'

print(
    f"fine-tune ID: {MODEL_ID}, "
    f"fine-tune status: {'STATUS_READY'}"
)

## Call Model

response = co.classify(
    inputs=[
        "i didnt feel humiliated",
        "i can go from feeling so hopeless to so damned hopeful just from being around someone who cares and is awake"
    ], model = MODEL_ID + "-ft"
)


print(response)

fine-tune ID: c62448a1-530a-45bb-8ca4-9e4950f39ac0, fine-tune status: STATUS_READY
id='c195e965-4725-4320-8149-60fb4163ed9c' classifications=[ClassifyResponseClassificationsItem(id='461255de-2ae9-4798-9514-2a8a2781f1bc', input='i didnt feel humiliated', prediction='sadness', predictions=['sadness'], confidence=0.9995045, confidences=[0.9995045], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=0.0003725919), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=4.0006358e-05), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=3.3491913e-05), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=4.2337746e-05), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=0.9995045), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=7.068441e-06)}, classification_type='single-label'), ClassifyResponseClassificationsItem(id='c6778b13-7f8a-425a-8ed3-965ef491d962', input='i can go from feeling so ho

fine-tune ID: c62448a1-530a-45bb-8ca4-9e4950f39ac0, fine-tune status: STATUS_READY
id='c195e965-4725-4320-8149-60fb4163ed9c' classifications=[ClassifyResponseClassificationsItem(id='461255de-2ae9-4798-9514-2a8a2781f1bc', 

input='i didnt feel humiliated', prediction='sadness', predictions=['sadness'], confidence=0.9995045, confidences=[0.9995045], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=0.0003725919), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=4.0006358e-05), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=3.3491913e-05), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=4.2337746e-05), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=0.9995045), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=7.068441e-06)}, classification_type='single-label'), ClassifyResponseClassificationsItem(id='c6778b13-7f8a-425a-8ed3-965ef491d962', 

input='i can go from feeling so hopeless to so damned hopeful just from being around someone who cares and is awake', prediction='sadness', predictions=['sadness'], confidence=0.99978495, confidences=[0.99978495], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=3.799905e-05), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=0.00010636302), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=2.0181142e-05), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=4.4122735e-05), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=0.99978495), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=6.412596e-06)}, classification_type='single-label')] meta=ApiMeta(api_version=ApiMetaApiVersion(version='1', is_deprecated=None, is_experimental=None), billed_units=ApiMetaBilledUnits(images=None, input_tokens=None, output_tokens=None, search_units=None, classifications=2.0), tokens=None, warnings=None)

In [12]:
# Convert Data
data_pth = "data/test.txt"
df = pd.read_csv(data_pth, sep=";", header=None, names=['text', 'label'])
df.to_csv("data/transformed-test.csv", index=False)

df.head()

text    label
0  im feeling rather rotten so im not very ambiti...  sadness
1          im updating my blog because i feel shitty  sadness
2  i never make her separate from me because i do...  sadness
3  i left with my bouquet of red and yellow tulip...      joy
4    i was feeling a little vain when i did this one  sadness

In [17]:
sample_output = ', '.join([f'"{i}"' for i in df.text.head(10)])
sample_output

'"im feeling rather rotten so im not very ambitious right now", "im updating my blog because i feel shitty", "i never make her separate from me because i don t ever want her to feel like i m ashamed with her", "i left with my bouquet of red and yellow tulips under my arm feeling slightly more optimistic than when i arrived", "i was feeling a little vain when i did this one", "i cant walk into a shop anywhere where i do not feel uncomfortable", "i felt anger when at the end of a telephone call", "i explain why i clung to a relationship with a boy who was in many ways immature and uncommitted despite the excitement i should have been feeling for getting accepted into the masters program at the university of virginia", "i like to have the same breathless feeling as a reader eager to see what will happen next", "i jest i feel grumpy tired and pre menstrual which i probably am but then again its only been a week and im about as fit as a walrus on vacation for the summer"'

In [43]:
from tqdm import tqdm

print(
    f"fine-tune ID: {MODEL_ID}, "
    f"fine-tune status: {'STATUS_READY'}"
)

## Call Model
batch_size = 90
all_results = []

num_batches = int(np.ceil(len(df) / batch_size))

for i in tqdm(range(num_batches)):
    start_idx = i * batch_size
    end_idx = min((i + 1) * batch_size, len(df))

    batch_texts = df.text[start_idx:end_idx].tolist()
    batch_labels = df.label[start_idx:end_idx].tolist()

    response = co.classify(
        inputs=batch_texts, 
        model = MODEL_ID + "-ft"
    )

    batch_results = [
        {
            'text': item.input, 
            'prediction': item.prediction, 
            'confidence': item.confidence,
            'true_label': batch_labels[j]  # true label directly
        }
        for j, item in enumerate(response.classifications)
    ]

    all_results.extend(batch_results)

# Create dataframe with predictions
# data = [
#     {
#         'text': item.input, 
#         'prediction': item.prediction, 
#         'confidence': item.confidence,
#         'true_label': test_labels[i]  # Add the true label directly
#     }
#     for i, item in enumerate(response.classifications)
# ]

df_output = pd.DataFrame(all_results)

# Display the results
accuracy = (df_output['prediction'] == df_output['true_label']).mean()
print(f"Overall accuracy: {accuracy:.4f}", "\n")

print("Sample Predictions", "\n")
print(df_output[['text', 'prediction', 'confidence', 'true_label']].head(10))

fine-tune ID: c62448a1-530a-45bb-8ca4-9e4950f39ac0, fine-tune status: STATUS_READY


100%|██████████| 23/23 [00:10<00:00,  2.18it/s]

Overall accuracy: 0.9320 

Sample Predictions 

                                                text prediction  confidence  \
0  im feeling rather rotten so im not very ambiti...    sadness    0.999861   
1          im updating my blog because i feel shitty    sadness    0.999815   
2  i never make her separate from me because i do...    sadness    0.999851   
3  i left with my bouquet of red and yellow tulip...        joy    0.999520   
4    i was feeling a little vain when i did this one    sadness    0.999912   
5  i cant walk into a shop anywhere where i do no...       fear    0.996856   
6   i felt anger when at the end of a telephone call      anger    0.990098   
7  i explain why i clung to a relationship with a...        joy    0.521207   
8  i like to have the same breathless feeling as ...        joy    0.998094   
9  i jest i feel grumpy tired and pre menstrual w...      anger    0.998516   

  true_label  
0    sadness  
1    sadness  
2    sadness  
3        joy  
4    sa

In [35]:
response = co.classify(
    inputs=["I give you a first look at what data analysis with Positron will look like"], model = MODEL_ID + "-ft"
)

data = [
    {
        'text': item.input, 
        'prediction': item.prediction, 
        'confidence': item.confidence,
        'true_label': test_labels[i]  # Add the true label directly
    }
    for i, item in enumerate(response.classifications)
]

output = pd.DataFrame(data)
print(output)

                                                text prediction  confidence  \
0  i give you a first look at what data analysis ...        joy    0.965547   

  true_label  
0    sadness  


----------------------------------------------------RAW RESPONSE TEXT----------------------------------------------------
fine-tune ID: c62448a1-530a-45bb-8ca4-9e4950f39ac0, fine-tune status: STATUS_READY
id='c79da695-ebdb-419d-a0f6-4d25c8f3307b' classifications=[ClassifyResponseClassificationsItem(id='dbec0e2d-1277-45ba-903b-216bfd639d44', input='" im feeling rather rotten so im not very ambitious right now "', prediction='sadness', predictions=['sadness'], confidence=0.9998542, confidences=[0.9998542], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=5.4011278e-05), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=2.6711423e-05), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=2.0701547e-05), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=3.9864328e-05), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=0.9998542), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=4.5297975e-06)}, classification_type='single-label'), ClassifyResponseClassificationsItem(id='974195b7-e004-402c-bc33-6bd824a5cd6f', input='" im updating my blog because i feel shitty "', prediction='sadness', predictions=['sadness'], confidence=0.9998038, confidences=[0.9998038], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=9.2773225e-05), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=4.2103e-05), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=3.5350564e-05), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=2.227356e-05), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=0.9998038), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=3.6969266e-06)}, classification_type='single-label'), ClassifyResponseClassificationsItem(id='4093b988-a1ac-4b8a-8aa1-48008f53ae1c', input='" i never make her separate from me because i don t ever want her to feel like i m ashamed with her "', prediction='sadness', predictions=['sadness'], confidence=0.99984527, confidences=[0.99984527], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=6.649932e-05), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=2.8157585e-05), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=2.080269e-05), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=3.470213e-05), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=0.99984527), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=4.5831525e-06)}, classification_type='single-label'), ClassifyResponseClassificationsItem(id='3d15f4c6-c033-4527-95f8-6883e273b9fc', input='" i left with my bouquet of red and yellow tulips under my arm feeling slightly more optimistic than when i arrived "', prediction='joy', predictions=['joy'], confidence=0.99957085, confidences=[0.99957085], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=2.0733048e-06), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=5.8212772e-05), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=0.99957085), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=0.0003382788), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=2.2421116e-05), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=8.152172e-06)}, classification_type='single-label'), ClassifyResponseClassificationsItem(id='290b956c-f241-416b-9ba0-591209088495', input='" i was feeling a little vain when i did this one "', prediction='sadness', predictions=['sadness'], confidence=0.9999082, confidences=[0.9999082], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=2.7912947e-05), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=2.0926258e-05), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=2.1380709e-05), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=1.6847593e-05), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=0.9999082), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=4.728935e-06)}, classification_type='single-label'), ClassifyResponseClassificationsItem(id='9e4d7517-4756-48d7-ac57-e14f75d6b6d5', input='" i cant walk into a shop anywhere where i do not feel uncomfortable "', prediction='fear', predictions=['fear'], confidence=0.9972558, confidences=[0.9972558], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=0.00047780748), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=0.9972558), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=2.1386559e-05), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=9.2898976e-05), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=0.001842387), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=0.00030970311)}, classification_type='single-label'), ClassifyResponseClassificationsItem(id='aaf9f48a-278d-4715-b4c1-0f8a7cfbb81f', input='" i felt anger when at the end of a telephone call "', prediction='anger', predictions=['anger'], confidence=0.9892603, confidences=[0.9892603], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=0.9892603), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=0.0034328243), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=0.000763353), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=5.6996738e-05), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=0.0064290366), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=5.744377e-05)}, classification_type='single-label'), ClassifyResponseClassificationsItem(id='f3f65e7a-bfc8-457a-8077-b0dcadfef964', input='" i explain why i clung to a relationship with a boy who was in many ways immature and uncommitted despite the excitement i should have been feeling for getting accepted into the masters program at the university of virginia "', prediction='joy', predictions=['joy'], confidence=0.5513103, confidences=[0.5513103], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=7.754919e-05), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=2.360502e-05), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=0.5513103), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=0.44821212), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=0.0003599873), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=1.6414713e-05)}, classification_type='single-label'), ClassifyResponseClassificationsItem(id='5f3b046c-3992-4377-a5bd-93abc4e6305f', input='" i like to have the same breathless feeling as a reader eager to see what will happen next "', prediction='joy', predictions=['joy'], confidence=0.99803776, confidences=[0.99803776], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=2.6950858e-05), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=0.00017746631), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=0.99803776), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=0.0016969593), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=3.347544e-05), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=2.7375272e-05)}, classification_type='single-label'), ClassifyResponseClassificationsItem(id='118ebf0e-2909-4a3b-adc5-122a443f64e7', input='" i jest i feel grumpy tired and pre menstrual which i probably am but then again its only been a week and im about as fit as a walrus on vacation for the summer "', prediction='anger', predictions=['anger'], confidence=0.9985562, confidences=[0.9985562], labels={'anger': ClassifyResponseClassificationsItemLabelsValue(confidence=0.9985562), 'fear': ClassifyResponseClassificationsItemLabelsValue(confidence=0.00078649103), 'joy': ClassifyResponseClassificationsItemLabelsValue(confidence=3.985148e-05), 'love': ClassifyResponseClassificationsItemLabelsValue(confidence=4.541623e-06), 'sadness': ClassifyResponseClassificationsItemLabelsValue(confidence=0.0005902062), 'surprise': ClassifyResponseClassificationsItemLabelsValue(confidence=2.2706687e-05)}, classification_type='single-label')] meta=ApiMeta(api_version=ApiMetaApiVersion(version='1', is_deprecated=None, is_experimental=None), billed_units=ApiMetaBilledUnits(images=None, input_tokens=None, output_tokens=None, search_units=None, classifications=10.0), tokens=None, warnings=None)